### ISSUE TITLE
TarReader stops enumerating entries when stream is unseekable

### ISSUE THREAD

**Issue by slang25** [dotnet/runtime#74728]
### Description

When enumerating a tar file using `TarReader` and `GetNextEntryAsync`, once `ExtractToFileAsync` is called on one entry then the enumerating stops.

However, if the source tar stream is copied into a `MemoryStream` then it works as expected.

### Reproduction Steps

```csharp
using System.Formats.Tar;
using static System.Console;

using System.IO.Compression;

// 🎸 Tar Issue that I wish you saw... 🎶

await using var s = File.OpenRead("./my-file.tar.gz");
await using var gz = new GZipStream(s, CompressionMode.Decompress, true);

// Change to true to see what should happen
var copyToMemoryStream = false;

var tarStream = copyToMemoryStream ? await CopyToMemoryStream(gz) : gz;
await using var tarReader = new TarReader(tarStream, true);

var temp = Directory.CreateDirectory("temp").FullName;

var entriesCount = 0;
while (await tarReader.GetNextEntryAsync() is { } entry)
{
	WriteLine(entry.Name);
	entriesCount++;
	
	var destination = Path.Combine(temp, entry.Name);
	if (entry.EntryType is TarEntryType.Directory)
	{
		Directory.CreateDirectory(destination);
	}
	else
	{
		await entry.ExtractToFileAsync(destination, true);
	}
}

WriteLine();
WriteLine("Should have encountered 140 tar entries");
WriteLine($"Actual: {entriesCount}");

async Task<Stream> CopyToMemoryStream(Stream source)
{
	var destination = new MemoryStream();
	await source.CopyToAsync(destination);
	destination.Seek(0, SeekOrigin.Begin);
	return destination;
}
```
[:octocat: Repro here](https://github.com/slang25/tar-issue)

### Expected behavior

In the code example provided, we should see the entries count being 140, and it indeed is when we copy the source to a `MemoryStream`.

### Actual behavior

The `TarReader` stops returning entries after the first one has been extracted, resulting in an entries count of 7 (we called `ExtractToFileAsync` on the 7th entry).

### Known Workarounds

Copy tar stream into a `MemoryStream` before using `TarReader`

### Other information

[:octocat: Repro here](https://github.com/slang25/tar-issue)

**Comment by danmoseley** [comment]
@carlossanlop

**Comment by am11** [comment]
This is fixed in the main branch and backported to release/7.0, but codeflow is waiting on https://github.com/dotnet/sdk/pull/27472t. You can build from source or download the latest runtime binaries (without SDK):

```sh
$ mkdir ~/.dotnet7_runtime && curl -sSL https://dotnetbuilds.blob.core.windows.net/public/Runtime/7.0.0-rc.2.22428.1/dotnet-runtime-7.0.0-rc.2.22428.1-linux-x64.tar.gz | tar xzf - -C ~/.dotnet7_runtime
$ alias dotnet7=~/.dotnet7_runtime

$ dotnet7 bin/Debug/net7.0/tar-issue.dll
...
Should have encountered 140 tar entries
Actual: 140
```

**Comment by danmoseley** [comment]
what fixed this? #72045?

**Comment by carlossanlop** [comment]
Hi @slang25, thanks for reporting.

This sounds like it's a repro of the bug we fixed in https://github.com/dotnet/runtime/pull/74396 which will be available in RC2. I'm going to double check anyway, using the latest bits in release/7.0.

Thanks for sharing the additional info, @am11.

**Comment by carlossanlop** [comment]
It works with the latest release/7.0 bits. The bug is fixed.

![image](https://user-images.githubusercontent.com/1175054/187112689-3f76b63b-8f91-4397-b4d2-486fc58dd985.png)

**Comment by carlossanlop** [comment]
Duplicate of https://github.com/dotnet/runtime/issues/74242

**Comment by danmoseley** [comment]
@slang25 Thanks for the report. We are interested in any issues you find in the latest bits.

**Comment by slang25** [comment]
Thanks, I tried with a daily build before, but yes the latest runtime has fixed it 🥳 

Very nice work, I'll keep an eye out for any other issues I bump into.
